README.md
# bggeb

bggen is the Hall D photoproduction code that Rakitha Beminiwattha <rakithab@jlab.org> imported for 
SoLID background simulations. The modifications were made to allow electroproduction event generation 
with this code. For now bggen can only produce events for proton target.

Jixie Zhang (jixie@jlab.org) changed it to be compiled and run in 64-bit mode. Jixie also made
some improvments as below.

1) remove Hall-D software dependence
2) self create random seed if given seed is zero
3) add beam current and target parameters into input file
4) add radiator thickness into input file
5) enable coherence brem. only if ratiator thickness > 1.0E-6
6) clean up code, add more comment and make it look consistant.

Check HISTORY.txt for details.

==============================================================
# To compile this program:

Note that CERN & ROOT must be installed before compilation.

## To compile use Makefile:

    make;
    bggen.Linux_x86_64.exe will be created.

## To compile use CMake:

    mkdir build && cd build
    cmake ../code
    make
    bggen will be created.
    
==============================================================

# To run this program:

All input files have been placed in run dir.

    cd run;
    ../bggen.Linux_x86_64.exe 

==============================================================

# To analyze the output:

    h2root bggen.nt bggen.root
    h2root bggen.hit bggen.his.root

    I happened to find that 'h2root' from root v5.28 or v6.10 and up can
    convert the hbook file into root file in the above. root v5.30 - v5.34
    can not.
     
    Use scripts/HallD_LH_xs.cc and scripts/HallD_gen_lund.cc to
    generate LUND files and kinematics plots.

    


HISTORY.txt
By Jixie Zhang @20180425
History of program bggen
Program bggen is actually from hall-d software package. The original introduction
of this program is in file README.HallD.
In 2015 Rakitha Beminiwattha extracted it out for moller and SoLID projects.
Check README.Rakitha for details about what he changed and how to compile|run it.
In April 2018, Jixie Zhang made modifications with Rakitha's version and created
version 1.0.0 and uploaded it to github. All the modifications are listed below.
For detail changes, please see file 'diff.txt'.


----------------------------------------------------------------------------
Version 1.1.1  20190716
1. Fixed bug in number of Z bin calculation. In previous version, if the given
    target thickness is very thin (< 0.001 rad.len.), the number of Z bin will be zero.
    This bug will cause the program crash since it trys to create 2-D histogram with zero
    bin number. This bug is now fixed in file brembeam_ini.F.
    
----------------------------------------------------------------------------
Version 1.1.0  20190502
1. Add Z dependence. For each Z bin, the window plus the target matetial in front of
   this Z bin will be used to create bremsstrahlung photon, while EPA is unchanged. 
   Therefore the downstream Z will see more bremsstrahlung photon than upstream.
   The Z bin width is determined as 0.5 cm or 0.1% radiation length, which is smaller.
   
A) bremsbeam_ini.F: modify 1D histo h1Egamma to h2EgammaZ, add Z loop
B) bg_eve.F: use 2-D histo 'h2EgammaZ' to throw z and Egamma
C) bg_ctrl.inc: Add variable 'NHZ',  which means number of Z bins in 'h2EgammaZ'
D) lowen_eve.F: Use histogram ID=9010 to determine bin index, before it is 9000
E) bg_ntup_ini.F and bg_evec.inc: Add variable vz into the commom block and ntuple output.
F) change hbook HLIMIT from 1.0E6 to 1.0E7 so it can take more records
G) change lrec from 1024 to 4096 for "CALL HROPEN(lun,'HISOUT','bggen.his','NQE',lrec,iost)"
H) change IQUEST(10) from 128000 to 12800000 so the output ntuple h9 can keep more records
I) fixed bug that SAID does not throw event. I have mistakes in filling histo h9010 before.
J) add this function: print out message "1000000 events simulated ..." every 1M events.
K) add run/go_bggen.csh to make it easier to run this program.

 
----------------------------------------------------------------------------
Version 1.0.0

1. Removed code/Makefile since it depends on Hall-D software.

2. Modified code/bg_hddm.c to disable HDDM related stuff.

3. Added Makefile which can compile this program in '-m64' mode.

4. Modified code/CMakeLists.txt to compile in '-m64' mode, the executable is linked
   against static library.
   
5. Added the following variables in input file:  
   beam current, target length, target density, target material.rad.length,
   and radiator thickness.
   From now on beam and target luminocity will be calculated using input parameters.
   Radiator thickness is used to control whether or not to use coherence bremsstrahlung
   or normal bremsstrahlung radiation.  If radiator thickness  > 1.0E-6 will 
   use coherence bremsstrahlung.
   
   The following files have been changed:
   code/bg_ctrl.inc
   code/bremsbeam_ini.F
   code/bg_ini.F
   code/cohbeam_ini.F
   code/cobrems.F

   These 3 cards were added in fort.15
C -- radiator thickness in cm, if >1.0E-6, will use coherence brem. photon
RADTHICK 0.00 
C -- electron beam current (uA)
CURRENT  15.0          
C -- Only Hydrogen target is supported, need more work for other target 
C -- tg length,density,material.rad.length. In cm, g/cm^3, g/cm^2.
TARGET  15.0   0.071   63.0470

6. Modified common block variable DCOLL (diameter of the radiator) to be in unit of cm.  
   Before it is in m, which is not consistant with other parameters in the same common block.
    
7. Changed bremsbeam_ini(xxx) and cohbeam_ini(xxx). Removed 'ZCOLLIM,COLDIAM'
   from argument list since they are inside common block.
    
8. Changed cobrems(xxx), add argument radThick.
    Subroutine cobrems(Emax,Epeak,colDist,colDiam,radThick)
    
9. Changed bremsbeam_ini(xxx). Added more comments. disable calling cobrems(xxx)
   since cobrems(xxx) only does initialization for coherence bremsstrahlung.
   It does nothing for normal bremsstrahlung radiation.    

10.Updated code/rnd_ini.F. If user given ZERO random seed in card 'RNDMSEQ', will
   use process ID and system time to form a seed, otherwise use the given see.
    


code/bg_ini.F
      SUBROUTINE BG_INI(IERR)
C
C---      Initialize the FFREAD and the relevant variables
C
      IMPLICIT NONE
      INTEGER IERR
C
      INCLUDE 'bg_ctrl.inc'
      INCLUDE 'bg_partc.inc'
C
      INTEGER mxffr,jffr
      PARAMETER (mxffr=10000)
      COMMON/CFREAD/ jffr(mxffr)
C
      INTEGER ier
C
      INTEGER i,j,lun,lout,iost,ip,kd(4),kf,lenc,idgea
      REAL am,wg
      CHARACTER cline*132
C
      lout=6
      IERR=1
C
      CALL HBOOK_INI
C
      NEVENT=0
      RUNNO=2
      IWROUT(1)=0
      IWROUT(2)=0
      IWROUT(3)=1
      IRND_SEQ=0
      NPRIEV=0
      EPH_LIM(1)=0.18
      EPH_LIM(2)=11 
      EELEC=11.0        ! electron energy
      EPEAK=8.          ! peak right edge
      ZCOLL=7600.       ! distance of radiator to collimator in cm
      DCOLL=0.34        ! diameter of collimator in cm 
      RADTHICK=0.0000   ! radiator thickness in cm (default is 0 um)
      EPYMIN=3.         ! min energy for PYTHIA
      ELOWMIN=0.15      ! minimal energy for the low energy generator 
C
      LUNWR(1)=0        ! HDDS file - LUN not used
      LUNWR(2)=2        ! sequential file
      LUNWR(3)=3        ! NTUPLE file
C
      VERTEX(1)=0.0     ! vertex x in cm
      VERTEX(2)=0.0     ! vertex y in cm
      VERTEX(3)=0.0     ! vertex z in cm
C
      CURRENT=15.0      ! beam current 15 uA
C     target parameter, added by Jixie @20180417, LH2 by default
      TGPARA(1)=15.0    ! target length cm
      TGPARA(2)=0.071   ! target density g/cm^3
      TGPARA(3)=63.047  ! target material rad. length in g/cm^2
      WINTHICK=0.000    ! target upstream window thickness in unit of rad. length
C
C---    Redefine FFREAD settings
C
      CALL FFINIT(mxffr)
      CALL FFSET('LINP',15)
      CALL FFSET('LOUT',6)
      CALL FFSET('SIZE',32)
      CALL FFSET('LENG',120)
C
      CALL FFKEY('TRIG'      , NEVENT     , 1,'INTEGER')
      CALL FFKEY('RUNNO'     , RUNNO      , 1,'INTEGER')
      CALL FFKEY('WROUT'     , IWROUT(1)  , 3,'INTEGER')
      CALL FFKEY('RNDMSEQ'   , IRND_SEQ   , 1,'INTEGER')
      CALL FFKEY('NPRIEV'    , NPRIEV     , 1,'INTEGER')
      CALL FFKEY('EPHLIM'    , EPH_LIM(1) , 2,'REAL')
      CALL FFKEY('EELEC'     , EELEC      , 1,'REAL')
      CALL FFKEY('EPEAK'     , EPEAK      , 1,'REAL')
      CALL FFKEY('ZCOLLIM'   , ZCOLL      , 1,'REAL')
      CALL FFKEY('DCOLLIM'   , DCOLL      , 1,'REAL')     
      CALL FFKEY('RADTHICK'  , RADTHICK   , 1,'REAL')    
      CALL FFKEY('EPYTHMIN'  , EPYMIN     , 1,'REAL')
      CALL FFKEY('ELOWMIN'   , ELOWMIN    , 1,'REAL')
      CALL FFKEY('VERTEX'    , VERTEX(1)  , 3,'REAL')
      CALL FFKEY('CURRENT'   , CURRENT    , 1,'REAL')
      CALL FFKEY('TARGET'    , TGPARA(1)  , 3,'REAL')
      CALL FFKEY('WINTHICK'  , WINTHICK   , 1,'REAL')
C
      CALL FFGO
C
C
C---   Read the particle masses (GEANT numbering)
C
      DO ip=1,MXPART
         AM_PART(ip)=0.
         WG_PART(ip)=0.
         DO i=1,4
            KD_PART(i,ip)=0
         ENDDO
      ENDDO
C
      lun=9
      OPEN(lun,FILE='particle.dat',STATUS='OLD',IOSTAT=iost
     +       ,FORM='FORMATTED')
      IF(iost.NE.0) THEN
         WRITE(lout,*) ' *** ERROR: Missing file particle.dat'
         GO TO 999
      ENDIF
 10   READ(lun,FMT='(A)',IOSTAT=iost) cline
      IF(iost.EQ.0) THEN
C
         IF(cline(1:1).NE.'*'.AND.cline(1:1).NE.'C') THEN
            READ(cline,*) ip,am,wg,kd
C            write(6,*) ip,am,wg,kd
            IF(ip.GT.0.AND.ip.LE.MXPART) THEN
               AM_PART(ip)=am
               WG_PART(ip)=wg
               DO i=1,4
                  KD_PART(i,ip)=kd(i)
               ENDDO
            ELSE
               WRITE(lout,*) ' --- ERROR: Reading file particle.dat ',
     +              'GEANT index is out of range ',ip 
            ENDIF
         ENDIF
C
         GO TO 10
C
      ELSE IF(iost.GT.0) THEN
         WRITE(lout,*) ' *** ERROR: Reading file particle.dat'
         GO TO 999
      ENDIF
C
C---   Read the GEANT<->PYTHIA particle table
C      
      DO i=1,MXPGEANT
         IPLUND(i)=0
         IDECLUND(i)=0
      ENDDO
      DO i=-MXPKC,MXPKC
         KCGEAN(i)=0
      ENDDO
C
      OPEN(lun,FILE='pythia-geant.map',STATUS='OLD',IOSTAT=iost
     +       ,FORM='FORMATTED')
      IF(iost.NE.0) THEN
         WRITE(lout,*) ' *** ERROR: Missing file pythia-geant.map'
         GO TO 999
      ENDIF
 15   READ(lun,'(A)',IOSTAT=iost) cline
      IF(iost.EQ.0) THEN
C
         lenc=0
         DO i=1,LEN_TRIM(cline)
            IF(cline(i:i).EQ.'!') GO TO 20
            lenc=i
         ENDDO
 20      CONTINUE
         IF(lenc.GE.3) THEN
            READ(cline(1:lenc),*) j,kf
            idgea=ABS(j)
         ENDIF
         IF(idgea.GT.0.AND.idgea.LE.MXPGEANT) THEN
            IF(kf.NE.0) THEN
               IPLUND(idgea) =kf
               IF(j.LT.0) IDECLUND(idgea)=1
            ENDIF
         ENDIF
C
         GO TO 15
C
      ELSE IF(iost.GT.0) THEN
         WRITE(lout,*) ' *** ERROR: Reading file pythia-geant.map'
         GO TO 999
      ENDIF
      CLOSE(lun)
C
      CALL RND_INI(IRND_SEQ)  ! random number initialization
C
      IF(EPH_LIM(1).LT.ELOWMIN) THEN
         WRITE(6,1005) ELOWMIN
 1005    FORMAT('  --- Initialization warning: EPH_LIM(1) is set'
     +         ,' to ELOWMIN:',F10.4)
         EPH_LIM(1)=ELOWMIN
      ENDIF
C
      IF(EPH_LIM(1).GT.EPH_LIM(2)) THEN
         WRITE(6,1000) EPH_LIM
 1000    FORMAT('  *** Initialization error: energy limits:',2F10.4)
         GO TO 999
      ELSE IF(EPH_LIM(1).EQ.EPH_LIM(2)) THEN
C
C---      Increase E2 slightly in order to make a valid histogram
C
         EPH_LIM(2)=EPH_LIM(1)*1.0001
C
      ELSE
C
C---       Bremsstrahlung beam: the E0 and Epeak should be cosistent
C
         IF(EELEC.LT.EPH_LIM(2)) THEN
            WRITE(6,1010) EELEC,EPH_LIM(2)
 1010       FORMAT('  *** Initialization error: Ee<E2:',2F10.4)
            GO TO 999
         ENDIF
C
         IF(EPEAK.GE.EELEC) THEN
            WRITE(6,1020) EPEAK,EELEC
 1020       FORMAT('  *** Initialization error: peak>Ee:',2F10.4)
            GO TO 999
         ENDIF
C
      ENDIF
C
C---  Photon Beam spectrum
C
      IDBEAM=9000
      NHBEA=0
      NHZ=0
!  use 1.0E-6 other than zero to avoid round off effect of fortran memory
      IF(RADTHICK.GT.1.0E-6) THEN
!  use coherant brems from diamond, in GeV or cm units
         CALL COHBEAM_INI(IDBEAM,EELEC,EPEAK,EPH_LIM)
      ELSE
!  use normal brem. radiation photon spectrum on proton target, in GeV or cm units
         CALL BREMSBEAM_INI(IDBEAM,EELEC,EPH_LIM)
      ENDIF
C
C---   Pythia
C
      IFPYTH=0
      IF(EPH_LIM(2).GT.EPYMIN) THEN
         CALL PYTH_INI(ier)
         IF(ier.NE.0) GO TO 999
         IFPYTH=1
      ENDIF
C
C---   Low energy processes
C
      IDLOWEN=0
      IF(EPH_LIM(1).LT.EPYMIN) THEN
         IDLOWEN=10000
         CALL LOWEN_INI(ier)
         IF(ier.NE.0) GO TO 999
      ENDIF
C
C---   Output file for HDDM
C
      IF(IWROUT(1).NE.0) THEN
         WRITE(6,*) ' WARNING: HDDM output disabled in this version!'
C         CALL OPEN_HDDM_OUTPUT('bggen.hddm')
      ENDIF
C
C---   Sequential output file
C
      IF(IWROUT(2).NE.0) THEN
         OPEN(LUNWR(2),FILE='bggen.dat',STATUS='UNKNOWN'
     +       ,FORM='UNFORMATTED')
      ENDIF
C
C---   NTUPLE
C
      IF(IWROUT(3).NE.0) THEN
         CALL BG_NTUP_INI(ier)
         IF(ier.NE.0) GO TO 999
      ENDIF
C
      IERR=0
 999  RETURN
      END

      
